pilot/pkg/model/push_context.go

```diff
@@ -347,6 +350,74 @@ func (ps *PushContext) UpdateMetrics() {
 	}
 }
 
+func boundGateways(env *Environment, proxyInstances []*ServiceInstance) []Config {
+	// collect workload labels
+	var workloadLabels LabelsCollection
+	for _, w := range proxyInstances {
+		workloadLabels = append(workloadLabels, w.Labels)
+	}
+
+	return env.Gateways(workloadLabels)
+}
+
+// FilteredGatewayServices filters the given set of services by associated hosts.
+// The association is proxy_instance --> gateways --> virtual services --> hosts.
+func (ps *PushContext) FilteredGatewayServices(proxy *Proxy, svcs []*Service) []*Service {
+
+	if !pilot.FilterGatewayServices {
+		return svcs
+	}
+
+	// gateway set.
+	gwmap := map[string]bool{}
+	// host set.
+	hm := map[string]bool{}
+
+	for _, gw := range boundGateways(ps.Env, proxy.ServiceInstances) {
+		gwmap[gw.Namespace+"/"+gw.Name] = true
+	}
+	log.Infof("FilteredGatewayServices: gateway %v has following gw resources:%v", proxy.ID, gwmap)
+
+	for _, vsc := range ps.VirtualServices(proxy, gwmap) {
+		vs, ok := vsc.Spec.(*networking.VirtualService)
+		if !ok { // should never happen
+			log.Errorf("in not a virtual service: %#v", vsc.Labels)
+			return svcs
+		}
+
+		for _, h := range vs.Http {
+			for _, r := range h.Route {
+				hm[r.Destination.Host] = true
+			}
+		}
+
+		for _, h := range vs.Tcp {
+			for _, r := range h.Route {
+				hm[r.Destination.Host] = true
+			}
+		}
+	}
+
+	log.Infof("FilteredGatewayServices: gateway %v has following hosts:%v", proxy.ID, hm)
+
+	filtered := make([]*Service, 0, len(svcs))
+
+	for _, s := range svcs {
+		svcHost := string(s.Hostname)
+
+		// explicitly associated with gateway or hidden association thru an istio component like
+		// istio-telemetry.ns.svc.cluster.local
+		// There has to be a better way to do this.
+		if hm[svcHost] || strings.HasPrefix(svcHost, "istio") || strings.Contains(svcHost, "kube-system") {
+			filtered = append(filtered, s)
```

Comment:
Lol. How is this going to scale when someone adds their own envoy access logs service (that flows in through a cluster).

Or adds their external authz service that points to some other cluster that’s from a service you are rejecting here? Same goes for all the authN policies (jwt authn, etc. that talk to an external service). This also breaks lua if it’s using a cluster.

Reply:
We explicitly model these in the gateway resource. These are hidden dependencies.
We know what is needed based on config, and hidden deps should be externalized.

Reply:
Yes but they don’t materialize in virtual service. A cluster referred to by envoyfilter may be a service in the mesh, but not exposed through gateway+virtual service. How do you tackle that ?

Reply:
We can add that to the envoy filter resource.
At present this resource has undeclared dependencies.
We make them explicit by adding 
‘Repeated string hosts’  . That we can exactly know what to keep.

I have updated description of this PR, check it out.